### PR TITLE
{mcrypt,libmcrypt}: fix build on darwin

### PR DESCRIPTION
**Pull request by szlend**
## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by szlend** [comment]
TODO: Tag with `backport release-23.11` once reviewed.

File: pkgs/development/libraries/libmcrypt/default.nix [hunk 1]
```diff
@@ -18,6 +18,10 @@ stdenv.mkDerivation rec {
       "ac_cv_func_realloc_0_nonnull=yes"
     ];
 
+    env = lib.optionalAttrs stdenv.cc.isClang {
+      NIX_CFLAGS_COMPILE = "-Wno-implicit-function-declaration -Wno-implicit-int";
+    };
```
Comment (sikmir) [inline pkgs/development/libraries/libmcrypt/default.nix:23]:
```suggestion
    env.NIX_CFLAGS_COMPILE = lib.optionalString stdenv.cc.isClang "-Wno-implicit-function-declaration -Wno-implicit-int";
```

**Comment by sikmir** [comment]
Result of `nixpkgs-review pr 269600` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>mcrypt</li>
    <li>steghide</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>libmcrypt</li>
  </ul>
</details>

**Comment by szlend** [comment]
@sikmir Thanks for the review, though those two packages were already broken. I don't think this PR be fixing downstream consumers should it?

File: pkgs/development/libraries/libmcrypt/default.nix (hunk 1 above)
Comment (szlend) [reply pkgs/development/libraries/libmcrypt/default.nix:23]: What's the advantage of doing it that way? With `optionalAttrs` you don't set `NIX_CFLAGS_COMPILE` to an empty string.

Though I just noticed there's an indentation error there so lemme fix that real quick.

Comment (sikmir) [reply pkgs/development/libraries/libmcrypt/default.nix]: Just more short form, I guess empty string is not a problem (or not?). But no problem, up to you)
Out of curiosity:
```sh
$ ag -c "env = lib.optionalAttrs stdenv.cc.isClang" | wc -l
      34
$ ag -c "NIX_CFLAGS_COMPILE = lib.optionalString stdenv.cc.isClang" | wc -l
      28
```

Comment (szlend) [reply pkgs/development/libraries/libmcrypt/default.nix]: The only difference I can think of is that `optionalAttrs` doesn't invalidate existing Linux builds unnecessarily. I think.

**Comment by sikmir** [comment]
Result of `nixpkgs-review pr 269600` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>steghide</li>
    <li>stegseek</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>libmcrypt</li>
    <li>mcrypt</li>
  </ul>
</details>

**Review by sikmir — APPROVED** [review]

**Pull request merged**